pkg/envoy/lds/listener.go

```diff
@@ -200,3 +150,148 @@ func getCIDRRange(cidr string) (*xds_core.CidrRange, error) {
+
+		marshalledFilter, err = envoy.MessageToAny(
+			getHTTPConnectionManager(route.OutboundRouteConfigName, cfg))
+		if err != nil {
+			log.Error().Err(err).Msgf("Error marshalling HTTPConnManager object")
+			return "", nil, err
+		}
+	}
+
+	return wellknownFilterName, marshalledFilter, nil
+}
+
+// getFilterChainMatchForService builds a filter chain to match the destination traffic.
+// Filter Chain currently match on destination IP for possible service endpoints
+func getFilterChainMatchForService(meshSvc service.MeshService, catalog catalog.MeshCataloger, cfg configurator.Configurator) (*xds_listener.FilterChainMatch, error) {
+	ret := &xds_listener.FilterChainMatch{}
+
+	endpoints, err := catalog.GetResolvableServiceEndpoints(meshSvc)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error getting GetResolvableServiceEndpoints for %s", meshSvc.String())
+		return nil, err
+	}
+
+	if len(endpoints) == 0 {
+		log.Error().Err(err).Msgf("No resolvable addresses retured for service %s", meshSvc.String())
+		return nil, errNoValidTargetEndpoints
+	}
+
+	for _, endp := range endpoints {
+		ret.PrefixRanges = append(ret.PrefixRanges, &xds_core.CidrRange{
+			AddressPrefix: endp.IP.String(),
+			PrefixLen: &wrapperspb.UInt32Value{
+				Value: singleIpv4Mask,
+			},
+		})
+	}
+
+	return ret, nil
+}
+
+func getFilterChains(catalog catalog.MeshCataloger, cfg configurator.Configurator, localSvc []service.MeshService) ([]*xds_listener.FilterChain, error) {
+	var ret []*xds_listener.FilterChain
+	var dstServicesSet map[service.MeshService]struct{} = make(map[service.MeshService]struct{}) // Set, avoid dups
+
+	outboundSvc, err := catalog.ListAllowedOutboundServices(localSvc[0])
+	if err != nil {
+		log.Error().Err(err).Msgf("Error getting allowed outbound services for %s", localSvc[0].String())
+		return nil, err
+	}
+
+	// Transform into set, when listing apex services we might face repetitions
+	for _, meshSvc := range outboundSvc {
+		dstServicesSet[meshSvc] = struct{}{}
+	}
+
+	// Getting apex services referring to the outbound services
+	splitServices := catalog.GetSMISpec().ListTrafficSplitServices()
+	for _, svc := range splitServices {
+		for _, outSvc := range outboundSvc {
+			if svc.Service == outSvc {
+				rootServiceName := kubernetes.GetServiceFromHostname(svc.RootService)
+				rootMeshService := service.MeshService{
+					Namespace: outSvc.Namespace,
+					Name:      rootServiceName,
+				}
+
+				// Add this root service into the set
+				dstServicesSet[rootMeshService] = struct{}{}
+			}
+		}
+	}
+
+	// Iterate all destination services
+	for keyService := range dstServicesSet {
+
+		// Filterchain for current dest
+		filterChain := xds_listener.FilterChain{
+			Name:             keyService.String(),
+			Filters:          []*xds_listener.Filter{},
+			FilterChainMatch: &xds_listener.FilterChainMatch{},
+		}
+
+		// Get filter for service
+		filterName, filterProto, err := getFilterForService(keyService, cfg)
```

Comment:
I can make this return the filter object filled already, instead of the inner elements